service/frontend/adminHandler.go

```diff
@@ -421,6 +435,249 @@ func (adh *AdminHandler) GetWorkflowExecutionRawHistory(
+	size := rawHistoryResponse.Size
+	// N.B. - Dual emit is required here so that we can see aggregate timer stats across all
+	// domains along with the individual domains stats
+	adh.metricsClient.RecordTimer(scope, metrics.HistorySize, time.Duration(size))
+	domainScope := adh.metricsClient.Scope(scope, metrics.DomainTag(request.GetDomain()))
+	domainScope.RecordTimer(metrics.HistorySize, time.Duration(size))
+
+	rawBlobs := rawHistoryResponse.HistoryEventBlobs
+	blobs := []*gen.DataBlob{}
+	for _, blob := range rawBlobs {
+		blobs = append(blobs, blob.ToThrift())
+	}
+
+	result := &admin.GetWorkflowExecutionRawHistoryV2Response{
+		HistoryBatches: blobs,
+		VersionHistory: versionHistory.ToThrift(),
+	}
+	if len(token.PersistenceToken) == 0 {
+		result.NextPageToken = nil
+	} else {
+		result.NextPageToken, err = serializeRawHistoryToken(token)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+func (adh *AdminHandler) validateGetWorkflowExecutionRawHistoryV2Request(
+	request *admin.GetWorkflowExecutionRawHistoryV2Request,
+) error {
+
+	execution := request.Execution
+	if len(execution.GetWorkflowId()) == 0 {
+		return &gen.BadRequestError{Message: "Invalid WorkflowID."}
+	}
+	// TODO currently, this API is only going to be used by re-send history events
+	// to remote cluster if kafka is lossy again, in the future, this API can be used
+	// by CLI and client, then empty runID (meaning the current workflow) should be allowed
+	if len(execution.GetRunId()) == 0 || uuid.Parse(execution.GetRunId()) == nil {
+		return &gen.BadRequestError{Message: "Invalid RunID."}
+	}
+
+	pageSize := int(request.GetMaximumPageSize())
+	if pageSize < 0 {
+		return &gen.BadRequestError{Message: "Invalid PageSize."}
+	}
+
+	if request.StartEventId == nil &&
+		request.StartEventVersion == nil &&
+		request.EndEventId == nil &&
+		request.EndEventVersion == nil {
+		return &gen.BadRequestError{Message: "Invalid event query range."}
+	}
+
+	if (request.StartEventId != nil && request.StartEventVersion == nil) ||
+		(request.StartEventId == nil && request.StartEventVersion != nil) {
+		return &gen.BadRequestError{Message: "Invalid start event id and start event version combination."}
+	}
+
+	if (request.EndEventId != nil && request.EndEventVersion == nil) ||
+		(request.EndEventId == nil && request.EndEventVersion != nil) {
+		return &gen.BadRequestError{Message: "Invalid end event id and end event version combination."}
+	}
+	return nil
+}
+
+func (adh *AdminHandler) updateEventRange(
+	request *admin.GetWorkflowExecutionRawHistoryV2Request,
+	versionHistories *persistence.VersionHistories,
+) (*persistence.VersionHistory, error) {
+
+	targetBranch, err := versionHistories.GetCurrentVersionHistory()
+	if request.StartEventId == nil || request.StartEventVersion == nil {
+		firstItem, err := targetBranch.GetFirstItem()
+		if err != nil {
+			return nil, err
+		}
+		request.StartEventId = common.Int64Ptr(common.FirstEventID)
```

Comment:
`request.StartEventId = common.Int64Ptr(common.FirstEventID) - 1`

Reply:
Need one special handling with the ContiansItem func in versionhistory